src/DataProtection/DataProtection/src/Internal/HostingApplicationDiscriminator.cs

```diff
@@ -21,6 +23,20 @@ public HostingApplicationDiscriminator(IHostEnvironment hosting)
         _hosting = hosting;
     }
 
-    // Note: ContentRootPath behavior depends on the version, sometimes it has a trailing slash, we normalize by default by removing a trailing slash
-    public string? Discriminator => _hosting?.ContentRootPath?.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
+    // Note: ContentRootPath behavior depends on the version, sometimes it does not have a trailing slash,
+    // we normalize by adding a trailing slash for non whitespace content root paths so data protection
+    // works across versions
+    public string? Discriminator
+    {
+        get
+        {
+            var contentRoot = _hosting?.ContentRootPath?.Trim();
+            if (string.IsNullOrEmpty(contentRoot) ||
+                contentRoot.EndsWith(DirectorySeparator, StringComparison.OrdinalIgnoreCase))
```

Comment:
Should we also check if it ends with Path.AltDirectorySeparatorChar?

Reply:
I had that in the first iteration but that leads to some weird behavior, right, since if it ends with a '/' it won't append, but it will fail since it won't match on windows.  Like we will never generate a content root path ourselves by default with the AltDirectorySeparator right?  Windows will always use \ and non windows will use / ?

Reply:
I recommended it because people often specify the wrong slash when deploying their app cross platform. E.g. developing on windows and deploying to linux. The alt separator still works in most APIs, and is invalid to combine with the primary separator (right?).

E.g. all of these are valid and we wouldn't want to add an additional separator to any of them, regardless of platform:
C:\dir/
C:\dir\
C:/dir\
C:/dir/

Reply:
Hrm, I'll add some variants that explicitly mix these so we can capture this behavior in our tests

Reply:
Its really unfortunate that we can't just trim by default, because then we would in a much better spot, but 6.0 has the trailing slash by default so we are stuck :(

Reply:
> Its really unfortunate that we can't just trim by default, because then we would in a much better spot, but 6.0 has the trailing slash by default so we are stuck :(

![image](https://user-images.githubusercontent.com/649492/170789804-d3e5c706-d4be-43bb-a8a1-38179a03b863.png)

Reply:
Good enough...